tests/common/util.rs

```diff
@@ -1191,26 +1221,613 @@ impl UCommand {
+    fn new(output: tempfile::NamedTempFile) -> Self {
+        Self {
+            current_file: output.reopen().unwrap(),
+            output,
+        }
+    }
+
+    /// Tries to clone the file pointer.
+    pub fn try_clone(&mut self) -> io::Result<File> {
+        self.output.as_file().try_clone()
+    }
+
+    /// Returns the captured output as [`String`]. Subsequent calls to any of the other output methods
+    /// will operate on the subsequent output.
+    pub fn output(&mut self) -> String {
+        String::from_utf8(self.output_bytes()).unwrap()
+    }
+
+    /// Returns the exact amount of bytes as `String`. Subsequent calls to any of the other output
+    /// methods will operate on the subsequent output.
+    ///
+    /// # Important
+    ///
+    /// This method blocks indefinitely if the amount of bytes given by `size` cannot be read
+    pub fn output_exact(&mut self, size: usize) -> String {
+        String::from_utf8(self.output_exact_bytes(size)).unwrap()
+    }
+
+    /// Returns the captured output as bytes. Subsequent calls to any of the other output methods
+    /// will operate on the subsequent output.
+    pub fn output_bytes(&mut self) -> Vec<u8> {
+        let mut buffer = Vec::<u8>::new();
+        self.current_file.read_to_end(&mut buffer).unwrap();
+        buffer
+    }
+
+    /// Returns all captured output, so far. Subsequent calls to any of the other output methods
+    /// will operate on the subsequent output.
+    pub fn output_all_bytes(&mut self) -> Vec<u8> {
+        let mut buffer = Vec::<u8>::new();
+        let mut file = self.output.reopen().unwrap();
+
+        file.read_to_end(&mut buffer).unwrap();
+        self.current_file = file;
+
+        buffer
+    }
+
+    /// Returns the exact amount of bytes. Subsequent calls to any of the other output methods will
+    /// operate on the subsequent output.
+    ///
+    /// # Important
+    ///
+    /// This method blocks indefinitely if the amount of bytes given by `size` cannot be read
+    pub fn output_exact_bytes(&mut self, size: usize) -> Vec<u8> {
+        let mut buffer = vec![0; size];
+        self.current_file.read_exact(&mut buffer).unwrap();
+        buffer
+    }
+}
+
+impl Default for CapturedOutput {
+    fn default() -> Self {
+        let mut retries = 10;
+        let file = loop {
+            let file = Builder::new().rand_bytes(10).suffix(".out").tempfile();
+            if file.is_ok() || retries <= 0 {
+                break file.unwrap();
+            }
+            sleep(Duration::from_millis(100));
+            retries -= 1;
+        };
+        Self {
+            current_file: file.reopen().unwrap(),
+            output: file,
+        }
+    }
+}
+
+impl Drop for CapturedOutput {
+    fn drop(&mut self) {
+        let _ = remove_file(self.output.path());
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum AssertionMode {
+    All,
+    Current,
+    Exact(usize, usize),
+}
+pub struct UChildAssertion<'a> {
```

Comment:
At this point, I'm wondering if we still need this? It's always `.make_assertion()` followed by `with_all_output`, `with_exact_output` or `with_current_output`, so those could be methods directly on `UChild`, right? `UChildAssertion` could still be used internally if that makes things easier, but I think we can hide it.

There's also `make_assertion().is_alive`, which could maybe just be `assert_alive` on `UChild`?

Reply:
Oh I see we need something for `kill` to return too. This should be just all the output in my opinion, so the same as `child.make_assertion().with_all_output()`

Reply:
I disagree, here. I really don't want having the assertions within `UChild`. The assertions are separated from actions, are tidied up and just play nicely together in `UChildAssertion`. This makes writing and reading the tests more convenient and fun. It's also a good place for adding new assertions if needed.

> This should be just all the output in my opinion, so the same as child.make_assertion().with_all_output()

No, not if you've asserted `with_current_output` or similar before you've called `wait` or `kill`. You'll just have to assert the rest of the output since the last assertion and not all output again. Please believe me. This assertion pattern is very common in `tail` and the way it is right now is nice and intuitive to work with. Note, if you haven't used any assertions since the start of the process then you'll get all output after a wait!

I would prefer leaving it to the user of `UChild` what to do after calling `kill()`. 

Overall, we should not try to be too closed on the usage of the api. We can't see in advance all needs of the tests and of `UChild`. The usage of the api should be up to the user of it and how he wants/needs to write the tests instead of enforcing a specific way of usage.

Reply:
Okay, I understand the difference, but I think that difference might not be super clear in the API at the moment, but I don't have any ideas to improve it. Let's leave it like you've written it now.